File: 21553 - Sai Nandan D/Week 3/239_Sliding Window Maximum.py
from collections import deque

class Solution:    
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        result, q = [], deque([])
        
        for i, n in enumerate(nums):
            while q and n > nums[q[-1]]:
                q.pop()
            q.append(i)
            if q[0] == i-k: q.popleft()
            if i >= k-1: result.append(nums[q[0]])
                
        return result

File: 21561 - Suvam Tamang/Week 3/239_Sliding Window Maximum.java
class Solution {
   
    public int[] maxSlidingWindow(int[] nums, int k) {
        int [] result = new int [nums.length-k+1];
        if(nums.length==1)
            return nums;
        Deque<Integer> Qi = new LinkedList<Integer>();
        for(int i=0;i<nums.length;i++){
            while(!Qi.isEmpty() && Qi.peekLast() < nums[i])
                Qi.removeLast();
            Qi.offerLast(nums[i]);
            
            if(i>=(k-1)){
                result[i-k+1] = Qi.peekFirst();
                if(!Qi.isEmpty() && Qi.peekFirst() == nums[i-k+1])
                    Qi.removeFirst();
            }
        }
        return result;
    }
}
